### ISSUE TITLE
Jil Serializer breaks Orleans

### ISSUE THREAD

**Issue by ZOXEXIVO** [dotnet/orleans#5347]
Sometimes, when i reboot my server, Web docker container can't connect to Silo container with this errors:

```
System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.InteropServices.PInvoke, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

File name: 'System.Runtime.InteropServices.PInvoke, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.ModuleHandle.ResolveType(RuntimeModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   at System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
   at System.Reflection.CustomAttribute.FilterCustomAttributeRecord(CustomAttributeRecord caRecord, MetadataImport scope, Assembly& lastAptcaOkAssembly, RuntimeModule decoratedModule, MetadataToken decoratedToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, Object[] attributes, IList derivedAttributes, RuntimeType& attributeType, IRuntimeMethodInfo& ctor, Boolean& ctorHasParameters, Boolean& isVarArg)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType, Boolean mustBeInheritable, IList derivedAttributes)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeAssembly assembly, RuntimeType caType)
   at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType, Boolean inherit)
   at System.Reflection.CustomAttributeExtensions.GetCustomAttributes[T](Assembly element)
   at Orleans.ApplicationParts.AssemblyAttributeFeatureProvider`1.PopulateFeature(IEnumerable`1 parts, TFeature feature)
   at Orleans.ApplicationParts.ApplicationPartManager.PopulateFeature[TFeature](TFeature feature)
   at Orleans.ApplicationPartManagerExtensions.CreateAndPopulateFeature[TFeature](IApplicationPartManager applicationPartManager)
   at Orleans.Serialization.SerializationManager.RegisterSerializers(IApplicationPartManager applicationPartManager)
   at Orleans.ClientBuilder.Build()
   at xxxx.Web.Application.ServiceRegistration.<AddOrleans>g__CreateOrleansClient|2_0(<>c__DisplayClass2_0& ) in /app/src/xxx.Web/Application/Registration.cs:line 72
```

When i make docker restart for my Web container, it's connect and everything is ok. 
But this error very annoying.

Docker image: microsoft/dotnet:2.2-aspnetcore-runtime (but i see these errors at v2.1 )
Tiered compilation if off

Web connect code looks like standard (i agree, that it looks like khm, .... ):

```
private static void AddOrleans(this IServiceCollection services, ApplicationSettings settings)
        {
            services.AddSingleton(CreateOrleansClient());

            IClusterClient CreateOrleansClient()
            {
                while (true)
                {
                    try
                    {
                        var clientBuilder = new ClientBuilder()
                            .UseStaticClustering(new IPEndPoint(IPAddress.Parse(settings.Orleans.ClusterIp), settings.Orleans.GatewayPort))
                            .Configure<ClusterOptions>(options =>
                            {
                                options.ClusterId = "xxx-dev";
                                options.ServiceId = "xxx";
                            }).ConfigureLogging(logging => logging.AddSerilog());

                        var client = clientBuilder.Build();
                        client.Connect().Wait();

                        return client;
                    }
                    catch (Exception ex)
                    {
                        Thread.Sleep(1000);
                        ///logging
                    }
                }
            }
        }
```

**Comment by sergeybykov** [comment]
@ReubenBond Any ideas? Looks weird.

**Comment by ZOXEXIVO** [comment]
Web project Dockerfile: 
```
FROM microsoft/dotnet:2.2-aspnetcore-runtime AS runtime
WORKDIR /app
COPY --from=build /app/src/xxx.Web/out .
ENTRYPOINT ["dotnet", "xxx.Web.dll"]
```
Caused only on docker host starting time

**Comment by ReubenBond** [comment]
That's missing some steps - I assume the build is `publish`ing to out?
Does the same issue occur if using the -sdk base?

**Comment by ZOXEXIVO** [comment]
Here project with issue. 
With binary search, I found, that problem in Jil serializer, but if caused by Orleans
 
https://github.com/ZOXEXIVO/OrleansIssue/

Clone it and run "run.sh", than in Docker logs you can see errors with PInvoke

**Comment by ReubenBond** [comment]
When I change your sln to connect to a silo, it works fine:
![image](https://user-images.githubusercontent.com/203839/52442951-5000a480-2ad9-11e9-8b0c-7c1ad979fd6c.png)

I'm not sure those file load exceptions are anything to worry about, they ought to be swallowed. Otherwise, the particular code path your looking at is just calling `assembly.GetCustomAttributes(...)`. It would be scary if that was not available for your assembly.

To avoid that `GetCustomAttributes` call on a Jil assembly, you can use `.ConfigureApplicationParts(parts => parts.AddAssembly(typeof(SomeTypeInYourGrainInterfacesAssembly)))` on the client builder. That will opt-out of the default behavior (scanning every assembly) since you specify which assembly to examine.

**Comment by ZOXEXIVO** [comment]
Main problem here - when my Client container start before starting Silo container - this error will repeat forever and client never be connected. Helps only restart Client container

Thanks for ConfigureApplicationParts, i will try it

**Comment by ReubenBond** [comment]
Closing due to inactivity - I hope this issue is resolved. If not, we can open a new issue & link to this.
